File: segwit_tweak.go
package deposit_address

import (
	"crypto/sha256"
	"github.com/decred/dcrd/dcrec/secp256k1/v4"

	"github.com/pkg/errors"
)

type PublicKey = secp256k1.PublicKey
type ModNScalar = secp256k1.ModNScalar
type JacobianPoint = secp256k1.JacobianPoint

const (
	AuxDataSize    = 32
	TweakSize      = 32
	SegwitTweakTag = "SegwitTweak"
)

// TweakPublicKey Computes a tweak and applies it to the specified public key. The tweak
// slice supplied to this function must be exactly 32 bytes long.
//
// The tweak hash is computed as sha256(tag || tag || pk || tweak) where
// 'tag' is sha256(SegwitTweakTag), 'pk' is the 33-byte compressed SEC1
// serialization of the public key, and 'tweak' is the supplied byte slice.
//
// The tweak hash is converted to a secp256k1 scalar t, and the tweaked
// public key is computed as PK + t * G, where G is the canonical generator.
func TweakPublicKey(pk *PublicKey, tweak []byte) (*PublicKey, error) {
	tweakScalar, err := computeTweakValue(pk, tweak)
	if err != nil {
		return nil, err
	}

	res := tweakWithScalar(pk, tweakScalar)
	return res, nil
}

// compute a tweaked public key as PK + tweak * G, for G the generator
func tweakWithScalar(pk *PublicKey, tweak *ModNScalar) *PublicKey {
	var tweakPoint, pkJacobian, sumJacobian JacobianPoint

	// tweak * G
	secp256k1.ScalarBaseMultNonConst(tweak, &tweakPoint)

	// PK + tweak * G
	pk.AsJacobian(&pkJacobian)
	secp256k1.AddNonConst(&tweakPoint, &pkJacobian, &sumJacobian)

	// convert to PublicKey
	sumJacobian.ToAffine()
	return secp256k1.NewPublicKey(&sumJacobian.X, &sumJacobian.Y)
}

// compute the tweak scalar for a given public key from a byte string
// Note: the length of the tweak must be exactly 32 bytes
func computeTweakValue(pk *PublicKey, tweak []byte) (*ModNScalar, error) {
	if len(tweak) != TweakSize {
		return nil, errors.Errorf("wrong size for tweak (got %v, want %v)", len(tweak), TweakSize)
	}

	if pk == nil {
		return nil, errors.New("nil public key")
	}

	h := sha256.New()

	// first, compute the tag bytes, sha256(SegwitTweakTag)
	h.Write([]byte(SegwitTweakTag))
	tag := h.Sum(nil)

	// now compute sha256(tag || tag || pk || tweak)
	h.Reset()
	h.Write(tag)
	h.Write(tag)
	h.Write(pk.SerializeCompressed())
	h.Write(tweak)
	tweakBytes := h.Sum(nil)

	// convert to ModNScalar
	var result ModNScalar
	overflowed := result.SetByteSlice(tweakBytes)
	if overflowed {
		// this happens with probability 2^-128. in other words, it will never happen
		return nil, errors.Errorf("tweak value resulted in overflow")
	}

	return &result, nil
}

File: ts/src/segwit_tweak.ts
import { secp256k1 } from "@noble/curves/secp256k1";
import { segwitTweakHasher } from "./hasher";
import type { Hex } from "./deposit_addr";

/**
 * Compute a tweak and apply it to the specified public key. The tweak value
 * supplied to this function must be exactly 32 bytes long.
 *
 * The tweak hash is computed as sha256(tag || tag || pk_bytes || tweak), where
 * 'tag' is sha256("SegwitTweakTag"), 'pk_bytes' is the 33-byte compressed SEC1
 * serialization of the public key, and 'tweak' is the 32-byte array supplied
 * as an argument.
 *
 * The tweak hash is converted to a secp256k1 scalar t, and the tweaked public
 * key is computed as 'PK + t * G', where 'G' is the canonical generator.
 *
 * @param { Hex } pk The public key to tweak.
 * @param { Uint8Array } tweak The 32-byte tweak value.
 * @return { Uint8Array } The tweaked publick ye as compressed SEC1 bytes.
 */
export function tweakPublicKey(pk: Hex, tweak: Uint8Array): Uint8Array {
  const tweakScalar = computeTweakValue(pk, tweak);
  return tweakWithScalar(pk, tweakScalar);
}

/**
 * Tweak the provided public key with the provided scalar.
 *
 * @param { Hex } pk The public key to tweak, as either compressed or uncompressed SEC1 bytes or hex string.
 * @param { bigint } tweak The tweak to apply.
 * @return { Uint8Array } The tweaked public key as compressed SEC1 bytes.
 */
function tweakWithScalar(pk: Hex, tweak: bigint): Uint8Array {
  // tweak * G
  const tweakPoint = secp256k1.ProjectivePoint.BASE.multiply(tweak);

  // PK
  const pubKey = secp256k1.ProjectivePoint.fromHex(pk);

  // PK + tweak * G
  const tweakedPubKey = pubKey.add(tweakPoint);

  // return compressed key
  const COMPRESSED = true;
  return tweakedPubKey.toRawBytes(COMPRESSED);
}

/**
 * Compute the tweak scalar for a given public key from a byte string.
 *
 * NOTE: the length of `tweak` must be exactly 32 bytes, and should be
 * the output of a collision-resistant hash function.
 *
 * @param { HEX } pk The public key to which the tweak will be applied, as * either compressed or uncompressed SEC1 bytes or hex string.
 * @param { Uint8Array } tweak The 32-byte input to the tweak computation.
 * @return { bigint } The tweak scalar.
 */
function computeTweakValue(pk: Hex, tweak: Uint8Array): bigint {
  if (tweak.length != 32) {
    throw new Error("`tweak` value must have length exactly 32");
  }

  // compute the SEC1 compressed public key
  const COMPRESSED = true;
  const pk_bytes = secp256k1.ProjectivePoint.fromHex(pk).toRawBytes(COMPRESSED);
  if (pk_bytes.length != 33) {
    throw new Error(
      "Expected `pk_bytes` to be 33 bytes, got a different length. This should not happen.",
    );
  }

  //     hash := sha256(tag || tag || pk_bytes || tweak)
  // segwitTweakHasher returns a hasher instance with `tag || tag` already processed
  const hashOut = segwitTweakHasher().update(pk_bytes).update(tweak).digest();

  // convert hash to a scalar by interpreting as a big-endian integer. This
  // will error if `hashOut` is an integer bigger than the order of secp256k1,
  // which happens with probability roughly 2^-128 (i.e., it will never happen).
  return secp256k1.utils.normPrivateKeyToScalar(hashOut);
}
